### PR TITLE
[Feature]Add cugraph local storage support to DGL

### PR DESCRIPTION
**Pull request by VibhuJawa**
## Description

This PR closes https://github.com/dmlc/dgl/issues/4672 by untegrating cugraph local storage into DGL. 

Accuracy Results on https://github.com/dmlc/dgl/tree/master/examples/pytorch/rgcn-hetero by using the same defaults mentioned there so **hyper parameter tuning ins not done.**  but we are still in a percentage point of the avg runs . 

| dataset | Avg(cugraph, 18 runs) | Avg(DGL, 5 runs) |
|---------|-----------------------|------------------|
| aifb    | 93.81                 | 94.44            |
| bgs     | 90.04                 | 91.04            |
| mutag   | 66.17                 | 67.37            |

This PR enables cugraph storage by using duck-typing by following the `Extending DGL Graph Sampling Pipeline
` RFC (See [link](https://hackmd.io/6PJixJj7Ssy19I-N9CV1Ag?view))  so all one needs to do is to switch the graph object from `DGLGraph` object to `CuGraphStorage` object. 

See below:
```diff

+from dgl.contrib.cugraph.convert import cugraph_storage_from_heterograph
+cugraph_g = cugraph_storage_from_heterograph(dgl_g)

sampler = dgl.dataloading.NeighborSampler(
        [15, 10, 5], prefetch_node_feats=['feat'], prefetch_labels=['label'])

train_dataloader = dgl.dataloading.DataLoader(
- dgl_g, 
+ cugraph_g,
train_idx, 
sampler, 
device=device, 
batch_size=1024,
shuffle=True,
drop_last=False, 
num_workers=0)

```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Features Added
-  [x] cugraph Homogeneous Graph
-  [x]  cugraph Heterogeneous Graph
-  [x] Single GPU cugraph support
-  [x] Multi GPU cugraph support
-  [x] End to End Examples 
-  [x] from hetrogenous_graph

##  Tests: 
-  [x] Homogeneous Graph Sampling
-  [x] Heterogeneous Graph Sampling
-  [x] from hetrogenous_graph

## Out of scope:
<del> renumber_vertices_by_type added in API
<del>  to_hetrogenous_graph
<del> edge_subgraph
<del> subgraph

### CODE REVIEW

File: tests/cugraph/test_from_dgl_hetrograph.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        assert_same_num_nodes(gs, g)
+        try:
+            # TODO: Raise an upstream issue for this
+            # to triage first time failure
+            assert_same_node_feats(gs, g)
+        except ValueError:
+            assert_same_node_feats(gs, g)
+        assert_same_num_edges_can_etypes(gs, g)
+        assert_same_num_edges_etypes(gs, g)
+        assert_same_edge_feats(gs, g)
```
Comment (VibhuJawa) [inline tests/cugraph/test_from_dgl_hetrograph.py]: Apart from sampling everything is tested here.

File: tests/cugraph/test_from_dgl_hetrograph.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    ]
+    for graph_f in graph_fs:
+        for idxtype in [th.int32, th.int64]:
+            g = graph_f(idxtype)
+            gs = cugraph_storage_from_heterograph(g)
+            assert_same_num_nodes(gs, g)
+            try:
+                # TODO: Raise an upstream issue for this
+                # to triage first time failure
+                assert_same_node_feats(gs, g)
```
Comment (VibhuJawa) [inline tests/cugraph/test_from_dgl_hetrograph.py]: Fails due to a PG issue that i need to find a minimal example for.

**Comment by alexbarghi-nv** [comment]
Looks fine to me.

File: tests/cugraph/test_from_dgl_hetrograph.py (hunk 2 above)
Comment (VibhuJawa) [reply tests/cugraph/test_from_dgl_hetrograph.py]: Will go away after this is merged: https://github.com/rapidsai/cugraph/pull/2804

File: examples/cugraph-pytorch/cugraph-local/rgcn-hetero/model.py [hunk 3]
```diff
@@ -0,0 +1,569 @@
+"""RGCN layer implementation"""
+# File same as: https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/model.py
+from collections import defaultdict
```
Comment (VibhuJawa) [inline examples/cugraph-pytorch/cugraph-local/rgcn-hetero/model.py]: No need to review as same as https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/model.py

File: examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    if args.validation:
+        val_idx = train_idx[: len(train_idx) // 5]
+        train_idx = train_idx[len(train_idx) // 5 :]
+    else:
+        val_idx = train_idx
+    
+    #Only change done
+    if args.use_cugraph_storage:
+        g = cugraph_storage_from_heterograph(g)
+
```
Comment (VibhuJawa) [inline examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py]: Only change done from `cugraph_storage_from_heterograph`

File: examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+Paper: https://arxiv.org/abs/1703.06103
+Reference Code: https://github.com/tkipf/relational-gcn
+"""
+# Modified the file at:
+# https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/entity_classify_mb.py
+# To work with cugraph_storage
+# Only change is:
+# from dgl.contrib.cugraph.convert import cugraph_storage_from_heterograph
+# g = cugraph_storage_from_heterograph(g)
+
```
Comment (VibhuJawa) [inline examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py]: Please see above comments as that  is the only change from upstream :-)

File: examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py (hunk 4 above)
Comment (wangxiaoyunNV) [reply examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py]: ok

File: python/dgl/contrib/cugraph/convert.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import dgl
+from .utils.cugraph_conversion_utils import (
```
Comment (wangxiaoyunNV) [inline python/dgl/contrib/cugraph/convert.py:2]: When from DGL to cugraph, did we ignore the dangling nodes?

File: python/dgl/contrib/cugraph/utils/cugraph_conversion_utils.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        if not gs.single_gpu:
+            df = dask_cudf.from_cudf(df, npartitions=16)
+
+        gs.add_node_data(
+            df, "node_id", ntype=ntype, contains_vector_features=False
+        )
+    return gs
+
+
+def add_nodes_from_dgl_heteroGraph(
```
Comment (wangxiaoyunNV) [inline python/dgl/contrib/cugraph/utils/cugraph_conversion_utils.py:74]: Did we also consider adding the dangling nodes here?

**Review by wangxiaoyunNV — COMMENTED** [review]
In the conversion function, will the dangling nodes also be taken care of? We do not what to ignore them.

File: python/dgl/contrib/cugraph/convert.py (hunk 6 above)
Comment (VibhuJawa) [reply python/dgl/contrib/cugraph/convert.py:2]: Dangling nodes with associated features will be handled , Can you give an example of a case you think is not handled , cause i think they should all be handled .
